### PR TITLE
Handle missing endpoints from discovery

### PR DESCRIPTION
**Pull request by jt-nti**
For some reason the discovery PeersOfChannel function does not
return an endpoint for the local peer, which the gateway was
relying on

Rather than updating PeersOfChannel to return the endpoint when
there is one, to match the Peers function, this change updates
how the gateway identifies the local peer

Signed-off-by: James Taylor <jamest@uk.ibm.com>

### CODE REVIEW

File: internal/pkg/gateway/registry.go [hunk 1]
```diff
@@ -165,12 +166,22 @@ func (reg *registry) endorsersByOrg(channel string, chaincode string) map[string
 	defer reg.configLock.RUnlock()
 
 	for _, member := range members {
+		pkiid := member.PKIid
+		// TODO use member.PreferredEndpoint() or go back to member.Endpoint?
```
Comment (andrew-coleman) [inline internal/pkg/gateway/registry.go]: I think we should stick to member.PreferredEndpoint()

File: internal/pkg/gateway/registry.go [hunk 2]
```diff
@@ -165,12 +166,22 @@ func (reg *registry) endorsersByOrg(channel string, chaincode string) map[string
...
 		// find the endorser in the registry for this endpoint
 		var endorser *endorser
-		if endpoint == reg.localEndorser.address {
+		if bytes.Equal(pkiid, reg.localEndorser.pkiid) {
+			logger.Debugw("Found local endorser", "pkiid", pkiid)
 			endorser = reg.localEndorser
+		} else if endpoint == "" {
+			// TODO the endpoint may be empty because discovery!!
+			// include seperate check/warning or just fall through to else???
+			reg.logger.Warnf("Failed to find endorser: no endpoint from discovery")
```
Comment (andrew-coleman) [inline internal/pkg/gateway/registry.go]: Keep the warning in.  We'd need to understand why we had an endorser with no endpoint if it's not the local peer.  Might be worth logging the pkiid of this endorser.

File: internal/pkg/gateway/registry.go [hunk 3]
```diff
@@ -312,18 +323,19 @@ func (reg *registry) registerChannel(channel string) error {
 	peers := map[string]string{}
 	members := reg.discovery.PeersOfChannel(common.ChannelID(channel))
 	for _, member := range members {
+		// TODO why is this fragile? Make it less fragile?
```
Comment (andrew-coleman) [inline internal/pkg/gateway/registry.go]: Matt was never happy about this - but this code was copied from the discovery service

**Review by andrew-coleman — COMMENTED** [review]
I'm wondering if we should also add a unit test that demonstrates how this fixes the empty endpoint problem.

File: internal/pkg/gateway/registry.go (hunk 3 above)
Comment (jt-nti) [reply internal/pkg/gateway/registry.go]: I'll leave it as is for now then

**Review by andrew-coleman — DISMISSED** [review]

**Comment by denyeart** [comment]
@jt-nti Please add PR and commit message explaining the motivation. Is this a workaround for something missing from discovery?

File: integration/gateway/endorsing_orgs_test.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+			for _, o := range peers {
+				mspid := network.Organization(o.Organization).MSPID
+
+				submitCheckEndorsingOrgTransaction(ctx, gatewayClient, signingIdentity, mspid)
+			}
+		}
+	})
+})
+
+func submitCheckEndorsingOrgTransaction(ctx context.Context, client gateway.GatewayClient, signingIdentity *nwo.SigningIdentity, mspids ...string) {
```
Comment (denyeart) [inline integration/gateway/endorsing_orgs_test.go:124]: The function name is singular (Org) while the input allows for passing multiple orgs (mspids). This causes some confusion.

File: integration/gateway/endorsing_orgs_test.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+		for _, p := range peers {
+			conn := network.PeerClientConn(p)
+			defer conn.Close()
+			gatewayClient := gateway.NewGatewayClient(conn)
+			ctx, cancel := context.WithTimeout(context.Background(), network.EventuallyTimeout)
+			defer cancel()
+
+			signingIdentity := network.PeerUserSigner(p, "User1")
+			for _, o := range peers {
```
Comment (denyeart) [inline integration/gateway/endorsing_orgs_test.go:115]: It took me a while to figure out why you have an outer and an inner peers loop. I finally concluded that you wanted to test every peer's gateway with every peer as the endorsing org.
Would be good to indicate this intent in a comment prior to the loop so that people know what to expect.

**Review by denyeart — APPROVED** [review]
The majority of the PR relates to testing the specification of endorsing org, while the PR and commit message talk only of the local peer workaround.  This would have been better as two distinct PRs, or at least two separate commits in a single PR.

I won't hold it up with my minor comments however.

**Comment by jt-nti** [comment]
@denyeart it was the new tests which uncovered the bug :)
